numpy/_core/src/multiarray/unique.cpp

```diff
@@ -18,77 +19,229 @@ struct FinalAction {
-                data += stride;
-            }
-        } while (iternext(iter));
+    PyThreadState *_save1 = PyEval_SaveThread();
+
+    // number of elements in the input array
+    npy_intp isize = PyArray_SIZE(self);
+
+    // variables for the string
+    npy_intp itemsize = descr->elsize;
+    npy_intp num_chars = itemsize / sizeof(T);
+    auto hash = [num_chars](const T *value) -> size_t {
+        return str_hash(value, num_chars);
+    };
+    auto equal = [itemsize](const T *lhs, const T *rhs) -> bool {
+        return std::memcmp(lhs, rhs, itemsize) == 0;
+    };
+
+    // Reserve hashset capacity in advance to minimize reallocations.
+    std::unordered_set<T *, decltype(hash), decltype(equal)> hashset(
+        isize * 2, hash, equal
+    );
+
+    // Input array is one-dimensional, enabling efficient iteration using strides.
+    char *idata = PyArray_BYTES(self);
+    npy_intp istride = PyArray_STRIDES(self)[0];
+    for (npy_intp i = 0; i < isize; i++, idata += istride) {
+        hashset.insert((T *)idata);
     }
 
     npy_intp length = hashset.size();
 
+    PyEval_RestoreThread(_save1);
+    NPY_ALLOW_C_API;
+    // NumPy API calls and Python object manipulations require holding the GIL.
+    PyObject *res_obj = PyArray_NewFromDescr(
+        &PyArray_Type,
+        descr,
+        1, // ndim
+        &length, // shape
+        NULL, // strides
+        NULL, // data
+        // This flag is needed to be able to call .sort on it.
+        NPY_ARRAY_WRITEABLE, // flags
+        NULL // obj
+    );
+
+    if (res_obj == NULL) {
+        return NULL;
+    }
+    NPY_DISABLE_C_API;
+    PyThreadState *_save2 = PyEval_SaveThread();
+
+    char *odata = PyArray_BYTES((PyArrayObject *)res_obj);
+    npy_intp ostride = PyArray_STRIDES((PyArrayObject *)res_obj)[0];
+    // Output array is one-dimensional, enabling efficient iteration using strides.
+    for (auto it = hashset.begin(); it != hashset.end(); it++, odata += ostride) {
+        std::memcpy(odata, *it, itemsize);
+    }
+
+    PyEval_RestoreThread(_save2);
+    return res_obj;
+}
+
+static PyObject*
+unique_vstring(PyArrayObject *self)
+{
+    /*
+    * Returns a new NumPy array containing the unique values of the input array.
+    * This function uses hashing to identify uniqueness efficiently.
+    */
+    NPY_ALLOW_C_API_DEF;
     NPY_ALLOW_C_API;
     PyArray_Descr *descr = PyArray_DESCR(self);
+    // NumPy API calls and Python object manipulations require holding the GIL.
     Py_INCREF(descr);
+    NPY_DISABLE_C_API;
+
+    // release the GIL
+    PyThreadState *_save1 = PyEval_SaveThread();
+
+    // number of elements in the input array
+    npy_intp isize = PyArray_SIZE(self);
+
+    // variables for the vstring
+    npy_string_allocator *in_allocator = NpyString_acquire_allocator((PyArray_StringDTypeObject *)descr);
+    auto in_allocator_dealloc = finally([&]() {
+        NpyString_release_allocator(in_allocator);
+    });
```

Comment:
nice, might be worth adding C++ wrappers for the `NpyString` API somewhere in NumPy so this pattern can be used elsewhere.